File: BleModule/src/main/java/com/mpdc4gsr/ble/RequestBuilder.java
package com.mpdc4gsr.ble;

import com.mpdc4gsr.ble.callback.RequestCallback;

import java.util.UUID;

import com.mpdc4gsr.commons.observer.Observe;
import com.mpdc4gsr.commons.poster.RunOn;

/**
 * date: 2019/9/20 18:00
 * author: bichuanfeng
 */
public class RequestBuilder<T extends RequestCallback> {
    String tag;
    RequestType type;
    UUID service;
    UUID characteristic;
    UUID descriptor;
    Object value;
    int priority;
    RequestCallback callback;
    WriteOptions writeOptions;

    RequestBuilder(RequestType type) {
        this.type = type;
    }


    public RequestBuilder<T> setTag(String tag) {
        this.tag = tag;
        return this;
    }


    public RequestBuilder<T> setPriority(int priority) {
        this.priority = priority;
        return this;
    }


    public RequestBuilder<T> setCallback(T callback) {
        this.callback = callback;
        return this;
    }

    public Request build() {
        return new GenericRequest(this);
    }
}

File: BleModule/src/main/java/com/mpdc4gsr/ble/RequestBuilder.kt
package com.mpdc4gsr.ble

import com.mpdc4gsr.ble.callback.RequestCallback
import java.util.UUID

open class RequestBuilder<T : RequestCallback?> internal constructor(var type: RequestType?) {
    var tag: String? = null
    var service: UUID? = null
    var characteristic: UUID? = null
    var descriptor: UUID? = null
    var value: Any? = null
    var priority: Int = 0
    var callback: RequestCallback? = null
    var writeOptions: WriteOptions? = null

    open fun setTag(tag: String?): RequestBuilder<T?>? {
        this.tag = tag
        return this
    }

    open fun setPriority(priority: Int): RequestBuilder<T?>? {
        this.priority = priority
        return this
    }

    open fun setCallback(callback: T?): RequestBuilder<T?>? {
        this.callback = callback
        return this
    }

    fun build(): Request {
        return GenericRequest(this)
    }
}
